index.js

```diff
@@ -35,12 +35,19 @@ function transform (chunk, enc, cb) {
   } else {
     this[kLast] += this[kDecoder].write(chunk)
     list = this[kLast].split(this.matcher)
+
+    if (this.keepMatcher) {
+      let l = []
+      list.length % 2 && list.push('') // Add trailing empty string if odd elements
```

Comment:
is this needed? can we avoid pushing this spurious element?

Reply:
> is this needed? can we avoid pushing this spurious element?

My usage case needs to know whether there is a `matcher` at the end of the last line of input.  With the following data pattern where there isn't a newline character at the end of the last line, using split2, it is not possible to know this until after the last line is read which is when the `end` event is fired.

```text
line1\n
line2\n
last line
```
I need to know if the last line is terminated with a newline or not when I read it.  

I thought this PR was a simple implementation that may have greater utility than my own needs.  It also matches implementations in other languages such as perl that retain the input record separator (i.e. newline) at the end of the string when reading text input, and provides the [`chomp`](https://perlmaven.com/chomp) function to remove said record separator from strings if so desired.  

But I'm open to other suggestions/implementations that would fulfill my usage case.

Cheers,
Damien.

Reply:
I understand your requirement. I do not understand why there is a `list.push('')`.

Reply:
Sorry @mcollina - I misunderstood your question.

From [String#split](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#Description):

> If separator is a regular expression that contains capturing parentheses, then each time separator is matched, the results (including any undefined results) of the capturing parentheses are spliced into the output array.

So parsing the example:

```
line1\n
line2\n
last line
```
Using the following matcher regular expression:

```javascript
  fs.createReadStream(file)
    .pipe(split(/(\r?\n)/,{keepMatcher:true}))
    .on('data', function (line) {
      //each chunk now is a separate line, and retains the \r\n or \n at the end!
    })
```
Will yield the following array:

```json
[
  "line1",
  "\n",
  "line2",
  "\n",
  "last line"
]
```
This line:

```javascript
list.length % 2 && list.push('') // Add trailing empty string if odd elements
```

Will add an empty string to the array if there are uneven number of pairs, such as with our example data (i.e. the last line has no newline character).

Ensuring the array has an even number of elements will prevent an overflow on the subsequent line of code:

```javascript
      for (var i = 0; i < list.length; i += 2) push(l, list[i] + list[i + 1])
```
where it loops over the array in increments of 2 to rejoin the `matcher` elements with their lines.  The empty string added to the end of `list` will be inconsequentially added to `last line` in the final loop, rather than an `undefined` value.

Is there a better way?

D.